### PR TITLE
Refactors circuitry into a component

### PR DESCRIPTION
**Pull request by Garen7**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: Garen
code: refactors integrated circuits into a component
code: wallframe assemblies are now wallframes
code: drone assemblies are now mobs
/:cl:

- [ ] dnm until #39484 is merged

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

Things that act like mobs should be mobs and things that act like wall frames should be wall frames. In addition this allow someone to make more interesting circuit assemblies and the assemblies such as the pen assembly and pda assembly can be made into actual pens and pdas.

I tested almost everything but i still recommend test merging this.

### CODE REVIEW

File: code/game/data_huds.dm [hunk 1]
```diff
@@ -436,52 +436,47 @@
 /*~~~~~~~~~~~~
 	Circutry!
 ~~~~~~~~~~~~~*/
-/obj/item/electronic_assembly/proc/diag_hud_set_circuithealth(hide = FALSE)
+/atom/proc/diag_hud_set_circuithealth(percent, hide = FALSE)
```
Comment (Garen7) [inline code/game/data_huds.dm]: Is it an issue that this proc is on atom? I had to put it there since the component can be attached to any atom

Comment (ninjanomnom) [reply code/game/data_huds.dm]: You should be doing this code from inside the component

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 2]
```diff
@@ -720,103 +181,151 @@
 	icon_state = "setup_large_industrial"
 	desc = "It's a case, for building large electronics with. This one resembles some kind of industrial machinery."
 
-/obj/item/electronic_assembly/drone
+/********
+ * Mobs *
+ ********/
+/mob/living/integrated_drone
```
Comment (Garen7) [inline code/modules/integrated_electronics/core/assemblies.dm]: is this fine just being a mob/living? when I had it as a simple mob I had to use some really bad define stuff to make it work so I reverted it to this.

File: code/__DEFINES/integrated_electronics.dm [hunk 3]
```diff
@@ -6,6 +6,12 @@
 #define DATA_CHANNEL "data channel"
 #define PULSE_CHANNEL "pulse channel"
 
+// All base objects that the integrated_electronic component is attached to
+#define IC_ASSEMBLY_PATHS list( /obj/item/electronic_assembly, /obj/item/wallframe/integrated_screen, /mob/living/integrated_drone )
```
Comment (ninjanomnom) [inline code/__DEFINES/integrated_electronics.dm]: `GLOB` not a define

File: code/__DEFINES/integrated_electronics.dm [hunk 4]
```diff
@@ -6,6 +6,12 @@
...
+
+// Get the material costs
+#define IC_GET_COST(max_circuits, max_complexity) (round((max_complexity + max_circuits) / 4) * SScircuit.cost_multiplier)
```
Comment (ninjanomnom) [inline code/__DEFINES/integrated_electronics.dm]: This can just be a proc on SScircuit

File: code/game/data_huds.dm (hunk 1 above)
Comment (ninjanomnom) [inline code/game/data_huds.dm]: The component needs to handle all it's own functionality and interactions as much as possible except where that functionality would make more sense as a generic functionality of the object it's on. This is pretty clearly only for circuits so circuits should be handling it.

File: code/game/data_huds.dm [hunk 5]
```diff
@@ -436,52 +436,47 @@
...
 	var/icon/I = icon(icon, icon_state, dir)
 	holder.pixel_y = I.Height() - world.icon_size
 	if((!isturf(loc))||hide) //if not on the ground dont show overlay
 		holder.icon_state = null
 	else
-		holder.icon_state = "huddiag[RoundDiagBar(obj_integrity/max_integrity)]"
+		holder.icon_state = "huddiag[RoundDiagBar(percent)]"
 
-/obj/item/electronic_assembly/proc/diag_hud_set_circuitcell(hide = FALSE)
+/atom/proc/diag_hud_set_circuitcell(percent, hide = FALSE)
```
Comment (ninjanomnom) [inline code/game/data_huds.dm]: Same as above

File: code/game/data_huds.dm [hunk 6]
```diff
@@ -436,52 +436,47 @@
...
-		var/chargelvl = battery.charge/battery.maxcharge
-		holder.icon_state = "hudbatt[RoundDiagBar(chargelvl)]"
-	else
+	else if(percent == -1)
 		holder.icon_state = "hudnobatt"
+	else
+		holder.icon_state = "hudbatt[RoundDiagBar(percent)]"
 
-/obj/item/electronic_assembly/proc/diag_hud_set_circuitstat(hide = FALSE) //On, On and dangerous, or Off
+/atom/proc/diag_hud_set_circuitstat(percent, prefered_hud_icon, combat_circuits, hide = FALSE) //On, On and dangerous, or Off
```
Comment (ninjanomnom) [inline code/game/data_huds.dm]: Same as above

File: code/game/data_huds.dm [hunk 7]
```diff
@@ -436,52 +436,47 @@
...
-	else if(battery.charge == 0)
+	else if(percent <= 0)
 		holder.icon_state = "hudoffline"
 	else if(combat_circuits) //has a circuit that can harm people
 		holder.icon_state = prefered_hud_icon + "-red"
 	else //Bot is on and not dangerous
 		holder.icon_state = prefered_hud_icon
 
-/obj/item/electronic_assembly/proc/diag_hud_set_circuittracking(hide = FALSE)
+/atom/proc/diag_hud_set_circuittracking(long_range_circuits, hide = FALSE)
```
Comment (ninjanomnom) [inline code/game/data_huds.dm]: Same as above

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by Hate9** [comment]
I can't wait for this to be ready and committed. It looks great.

**Pull request closed without merging**